crates/uv/src/commands/tool/upgrade.rs

```diff
@@ -398,5 +448,44 @@ async fn upgrade_tool(
         )?;
     }
 
-    Ok(outcome)
+    let constraint = match &outcome {
+        UpgradeOutcome::UpgradeDependencies | UpgradeOutcome::NoOp => {
+            pinned_requirement_version(&existing_tool_receipt, name)
+                .map(|version| UpgradeConstraint::PinnedVersion { version })
+        }
+        UpgradeOutcome::UpgradeTool | UpgradeOutcome::UpgradeEnvironment => None,
+    };
+
+    Ok(UpgradeReport {
+        outcome,
+        constraint,
+    })
+}
+
+fn pinned_requirement_version(tool: &Tool, name: &PackageName) -> Option<Version> {
+    pinned_version_from(tool.requirements(), name)
+        .or_else(|| pinned_version_from(tool.constraints(), name))
+}
+
+fn pinned_version_from(requirements: &[Requirement], name: &PackageName) -> Option<Version> {
+    requirements
+        .iter()
+        .filter(|requirement| requirement.name == *name)
+        .find_map(|requirement| match &requirement.source {
+            RequirementSource::Registry { specifier, .. } => {
+                let mut specifiers = specifier.iter();
+
+                let first = specifiers.next()?;
+
+                if specifiers.next().is_some() {
+                    return None;
+                }
+
+                match first.operator() {
+                    Operator::Equal | Operator::ExactEqual => Some(first.version().clone()),
+                    _ => None,
+                }
```

Comment:
It'd be weird, but wouldn't this miss a pin mixed with other constraints, e.g., `>=0.1, ==0.2`? Would it make more sense to search for a ` Operator::Equal` in the set of specifiers?

Reply:
Ah yeah interesting, I was able to break it by specifying `babel>=2.0,==2.6.0`, although not sure you see this much in the wild 😅